site/blog/2022/nodejs-deploy-codepipeline.mdx

```diff
@@ -0,0 +1,221 @@
+---
+slug: deploy-nodejs-app-with-codepipeline
+title: Deploy a NodeJS App on AWS using CodePipeline
+date: 2022-12-23
+authors: [yrobla]
+---
+
+# Deploy a NodeJS app to AWS with one click using CodePipeline
+
+Codepipeline is a service provided by AWS to provide a fully automated Continuous delivery service. It allows to fully manage all the steps in the code delivery lifecycle, from checking the source code and building it, testing and deploying to production. The whole process is managed by CodePipeline using specific configuration files without manual intervention, offering a safe and reliable way to deploy to the cloud.
+
+The following example will showcase the automated build and deployment of a NodeJS app. The testing part is intentionally not covered by this scope, but would be an important part to cover when deploying to production.
+
+## The NodeJS application
+
+In this example, we are going to build and deploy a very simple NodeJS application, that simply will output **Hello World!** to the browser. The application will be served to port 80 by Nginx, that will be acting as a reverse proxy for the node app. The application is hosted on a public github repository: https://github.com/iasql/iasql-code-example-node
+
+The indications for building and deploying the whole application are defined in the `appspec.yml` file: https://github.com/iasql/iasql-code-example-node/blob/main/appspec.yml . This file contains all the information needed by CodeDeploy to manage the whole lifecycle of the application (Complete reference can be seen at https://docs.aws.amazon.com/codedeploy/latest/userguide/reference-appspec-file-example.html#appspec-file-example-server).
+
+This sample configuration is divided in two steps: **files** and **hooks**. 
+**Files** section allows to specify where to copy the contents of the application. The contents of this application are divided between main content of the app, and specific nginx configuration.
+**Hooks** section allows to define scripts to tackle every step of the app's lifecycle. In this case those steps are covered:
+- BeforeInstall: script to install all needed dependencies to run the application. It is used to install all packages and dependencies
+- AfterInstalll: script triggered after all the dependencies have been installed but before the app is receiving any traffic. It is the best place to perform additional installation steps of the application, such as `npm install`
```

Comment:
I don't get the mentioning of `npm install` here. Why is that not part of `BeforeInstall`? Based on the description there: "It is used to install all packages and dependencies" which is exactly what `npm install` does for a node process.

```suggestion
- AfterInstall: script triggered after all the dependencies have been installed but before the app is receiving any traffic. It is the best place to perform additional installation steps of the application, such as `npm install`
```